### DISCUSSION TITLE
What is the best way to insert class-style pages with dynamic content using a loop?

### DISCUSSION THREAD

**Discussion by jbengelsdorf** [ctreffe/alfred#111] [Alfred Q&A]
The [alfred3 documentation](https://alfredo3.psych.bio.uni-goettingen.de/docs/index.html) provides an excellent demonstration of [how to use loops](https://alfredo3.psych.bio.uni-goettingen.de/docs/howto_loops.html) for adding pages written in instance-style to a section written in class-style. Since I rely on the [page hooks](https://alfredo3.psych.bio.uni-goettingen.de/docs/howto_hooks.html) due to dynamic content on the pages, I am wondering how I can add pages written in class-style to a section using a loop.

**Comment by ctreffe** [comment]
The following minimal example shows how to use a loop to add multiple instances of a page created in class style to a section:

```python
import alfred3 as al

exp = al.Experiment()

# The following page will be repeatedly added to the experiment
class MyRepeatedPage(al.Page):
    def on_first_show(self):
        self += al.Text('I am the repeated page.')

# We use a standard section's on_exp_access hook to execute our loop
@exp.member
class MySection(al.Section):
    def on_exp_access(self):
        # In the loop, we create instances of our previously created page class
        for i in range(10):
            self.exp += MyRepeatedPage(title=f"Page {i+1}", name=f"repeated_{i+1}")

if __name__ == "__main__":
    exp.run()
```

__CAUTION__: When adding multiple instances of the same page to an experiment, you need to assign different names for each input element on each of these pages. There are two ways to achieve this.

As the preferred solution, you can use the argument [_prefix\_element\_names_](https://alfredo3.psych.bio.uni-goettingen.de/docs/generated/alfred3.page.Page.prefix_element_names.html) to automatically add the (already varying) page name as a prefix to all element names on the page. This is achieved with a minimal change in the example above:

```python
# We use a standard section's on_exp_access hook to execute our loop
@exp.member
class MySection(al.Section):
    def on_exp_access(self):
        # In the loop, we create instances of our previously created page class
        for i in range(10):
            self.exp += MyRepeatedPage(title=f"Page {i+1}", name=f"repeated_{i+1}", prefix_element_names = True)

```

As an alternative solution, you can use the page's [_vargs_ dictionary](https://alfredo3.psych.bio.uni-goettingen.de/docs/generated/alfred3.page.Page.vargs.html) to pass your counter (and any other variables) to the repeated page and subsequently use the counter when naming elements:

```python
import alfred3 as al

exp = al.Experiment()

# The following page will be repeatedly added to the experiment
class MyRepeatedPage(al.Page):
    def on_first_show(self):
        self += al.Text('I am the repeated page.')
        self += al.TextEntry(name=f"entry_{self.vargs.counter}")


# We use a standard section's on_exp_access hook to execute our loop
@exp.member
class MySection(al.Section):
    def on_exp_access(self):
        # In the loop, we create instances of our previously created page class
        for i in range(10):
            self.exp += MyRepeatedPage(
                title=f"Page {i+1}",
                name=f"repeated_{i+1}",
                vargs={"counter": i+1}
            )

if __name__ == "__main__":
    exp.run()
```
